### PR TITLE
Faster headers copy

### PR DESCRIPTION
**Pull request by louiscryan**
Improves performance of copying headers when the hashing strategy and name validator are exactly equivalent

Also fixes an issue in CombinedHttpHeaders when copying from another instance of the same type which caused double escaping

### CODE REVIEW

**Comment by louiscryan** [comment]
@nmittler @buchgr

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java [hunk 1]
```diff
@@ -64,6 +67,31 @@ public CharSequence escape(CharSequence value) {
...
 
     @Override
+    public CombinedHttpHeaders add(Headers<? extends CharSequence> headers) {
+        // Override the fast-copy mechanism used by DefaultHeaders
+        checkNotNull(headers, "headers");
+        if (headers == this) {
+            throw new IllegalArgumentException("can't add to itself.");
+        }
+        if (headers instanceof CombinedHttpHeaders) {
+            if (this.isEmpty()) {
```
Comment (buchgr) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: nit: get rid of `this`.

File: codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java [hunk 2]
```diff
@@ -148,6 +148,59 @@ public void testContains() {
...
+        assertFalse(headers.containsLong(bs("byte"), Byte.MIN_VALUE));
+
+        assertTrue(headers.containsDouble(bs("double"), Double.MAX_VALUE));
+        assertFalse(headers.containsDouble(bs("double"), Double.MIN_VALUE));
+
+        assertTrue(headers.containsFloat(bs("float"), Float.MAX_VALUE));
+        assertFalse(headers.containsFloat(bs("float"), Float.MIN_VALUE));
+
+        assertTrue(headers.containsTimeMillis(bs("millis"), millis));
+        // This test doesn't work on midnight, January 1, 1970 UTC
```
Comment (buchgr) [inline codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java]: that's a blocker :laughing:

**Comment by louiscryan** [comment]
@buchgr resolved merge conflicts etc. PTAL

**Comment by nmittler** [comment]
@louiscryan LGTM

@Scottmitch do you want to take a look as well?

**Comment by buchgr** [comment]
@louiscryan looks good. why/where do you care about the performance of this?

**Comment by louiscryan** [comment]
GRPC is going to store a header template per signature and use that when creating it's calls to avoid some of the costs we incur today converting to Netty headers.

These copying costs showed up in GRPC profiling. I'll provide some no's....

**Comment by louiscryan** [comment]
Here's the WIP in GRPC
https://github.com/grpc/grpc-java/pull/1183

**Comment by buchgr** [comment]
ic. makes sense :).

**Comment by louiscryan** [comment]
@Scottmitch FYI this PR can be backported to 4.1

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java [hunk 3]
```diff
@@ -64,6 +67,42 @@ public CharSequence escape(CharSequence value) {
     }
 
     @Override
+    public HttpHeaders add(Headers<? extends CharSequence, ? extends CharSequence, ?> headers) {
+        // Override the fast-copy mechanism used by DefaultHeaders
+        checkNotNull(headers, "headers");
+        if (headers == this) {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: duplicate null check?

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: consider moving the previous null check into `else` block of the `instanceof` check.

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java [hunk 4]
```diff
@@ -64,6 +67,42 @@ public CharSequence escape(CharSequence value) {
...
+                add(header.getKey(), header.getValue());
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public HttpHeaders set(Headers<? extends CharSequence, ? extends CharSequence, ?> headers) {
+        checkNotNull(headers, "headers");
+        if (headers == this) {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: duplicate null check?

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java [hunk 5]
```diff
@@ -64,6 +67,42 @@ public CharSequence escape(CharSequence value) {
...
+            for (Map.Entry<? extends CharSequence, ? extends CharSequence> header : headers) {
+                add(header.getKey(), header.getValue());
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public HttpHeaders set(Headers<? extends CharSequence, ? extends CharSequence, ?> headers) {
+        checkNotNull(headers, "headers");
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: null is already check in `add`. can we remove this?

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java [hunk 6]
```diff
@@ -64,6 +67,42 @@ public CharSequence escape(CharSequence value) {
...
+        if (headers == this) {
+            throw new IllegalArgumentException("can't add to itself.");
+        }
+        clear();
+        add(headers);
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: nit: `return add(headers);`

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 7]
```diff
@@ -379,21 +379,36 @@ public T add(Headers<? extends K, ? extends V, ?> headers) {
...
             @SuppressWarnings("unchecked")
-            DefaultHeaders<K, V, T> defaultHeaders = (DefaultHeaders<K, V, T>) headers;
-            HeaderEntry<K, V> e = defaultHeaders.head.after;
-            while (e != defaultHeaders.head) {
-                add(e.key, e.value);
-                e = e.after;
+            DefaultHeaders<? extends K, ? extends V, T> defaultHeaders =
+                    (DefaultHeaders<? extends K, ? extends V, T>) headers;
+            HeaderEntry<? extends K, ? extends V> e = defaultHeaders.head.after;
+            if (defaultHeaders.hashingStrategy == this.hashingStrategy &&
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: nit: remove `this.`

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 8]
```diff
@@ -542,17 +557,7 @@ public T set(Headers<? extends K, ? extends V, ?> headers) {
...
-            while (e != defaultHeaders.head) {
-                add(e.key, e.value);
-                e = e.after;
-            }
-        } else {
-            add(headers);
-        }
+        addImpl(headers);
         return thisT();
     }
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: can `setAll` also use `addImpl(headers);`?

File: codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java [hunk 9]
```diff
@@ -14,6 +14,7 @@
  */
 package io.netty.handler.codec;
 
+import com.google.protobuf.ByteString;
```
Comment (Scottmitch) [inline codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java]: Can you just use `AsciiString`?

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 10]
```diff
@@ -379,21 +379,36 @@ public T add(Headers<? extends K, ? extends V, ?> headers) {
...
+            HeaderEntry<? extends K, ? extends V> e = defaultHeaders.head.after;
+            if (defaultHeaders.hashingStrategy == this.hashingStrategy &&
+                    defaultHeaders.nameValidator == this.nameValidator) {
+                // Fastest copy
+                while (e != defaultHeaders.head) {
+                    add0(e.hash, index(e.hash), e.key, e.value);
+                    e = e.after;
+                }
+            } else {
+                // Fast copy
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: Do you have metrics on this path vs "slow" path?

**Comment by Scottmitch** [comment]
@louiscryan - In general LGTM. A few comments / questions

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java (hunk 5 above)
Comment (louiscryan) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: Would be a slight change in behavior as clear() would get called before null check. I'm happy either way

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java (hunk 6 above)
Comment (louiscryan) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: Done

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 7 above)
Comment (louiscryan) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: Done

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 8 above)
Comment (louiscryan) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: Done

File: codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java (hunk 9 above)
Comment (louiscryan) [inline codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java]: That's very odd. Removed

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 10 above)
Comment (louiscryan) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: Here's some for 'fastest' vs 'fast'. Adding benchmark to PR

Benchmark                         (exampleHeader)  Mode  Cnt    Score    Error  Units
HeadersBenchmark.http2AddAllFastest           SIX  avgt   20  149.917 ±  9.492  ns/op
HeadersBenchmark.http2AddAllFast              SIX  avgt   20  246.499 ± 18.825  ns/op
HeadersBenchmark.httpAddAllFastest               SIX  avgt   20  109.263 ± 13.840  ns/op

**Comment by louiscryan** [comment]
Note that the http header copy fastest path is decently faster than http2s which is a curiosity

**Comment by louiscryan** [comment]
Fixed checkstyle issue and nettified the commit message format

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java (hunk 10 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: You have 2 `http2AddAllFastest` labels and 1 `http2AddAllFast` in the text above. I don't see `http2AddAllSlow`. Can you clarify the benchmark names and association with numbers?

I'm guessing the numbers follow the `Fastest`, `Fast`, `Slow` terminology?

File: microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java [hunk 11]
```diff
@@ -139,4 +139,22 @@ public void http2Iterate(Blackhole bh) {
...
+    @Benchmark
+    @BenchmarkMode(Mode.AverageTime)
+    public void http2AddAllFast(Blackhole bh) {
+        bh.consume(new DefaultHttp2Headers(false).add(http2Headers));
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.AverageTime)
+    public void http2AddAllSlow(Blackhole bh) {
+        bh.consume(new DefaultHttp2Headers().add(http2Headers));
```
Comment (Scottmitch) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: @louiscryan - Can you re-run with validation off (to be fair to other benchmarks)? `new DefaultHttp2Headers(false)`

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: Not ideal to have to duplicate the preconditions of `add` in `set`. Based on the API docs I don't think we violate them by removing these precondition checks and letting `addImpl` enforce the checks. However if we can't do this (or don't want to), can we centralize the precondition logic into another utility method?

**Comment by buchgr** [comment]
```
Benchmark                         Mode  Cnt    Score    Error  Units
HeadersBenchmark.http2AddAllFast  avgt   10  243.314 ± 28.893  ns/op
HeadersBenchmark.http2AddAllSlow  avgt   10  242.381 ± 17.805  ns/op
HeadersBenchmark.httpAddAllFast   avgt   10  180.437 ± 11.777  ns/op
```

I am getting this for headers of size 11.

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java (hunk 5 above)
Comment (louiscryan) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: Is the checkNotNull really adding much value? In all these cases NPE will be thrown immediately as soon as it's used. checkNotNull is mostly useful for state initialization in an object but that's not what is happening here.

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: I agree. We should only have to check `null` if we aren't "dereferencing" in the same context.

**Comment by louiscryan** [comment]
I'm running OpenJDK7 on a linux box. With the latest commit which clearly separates fastest / fast / slow paths I get

```
Benchmark                            (exampleHeader)  Mode  Cnt     Score     Error  Units
HeadersBenchmark.http2AddAllFastest             FIVE  avgt   10   140.406 ±  17.154  ns/op
HeadersBenchmark.http2AddAllFast                FIVE  avgt   10   203.138 ±  32.585  ns/op
HeadersBenchmark.http2AddAllSlow                FIVE  avgt   10   244.247 ±  72.164  ns/op
HeadersBenchmark.http2AddAllFastest        TWENTYTWO  avgt   10   482.338 ±  16.192  ns/op
HeadersBenchmark.http2AddAllFast           TWENTYTWO  avgt   10  1182.494 ± 260.362  ns/op
HeadersBenchmark.http2AddAllSlow           TWENTYTWO  avgt   10  1032.513 ± 221.226  ns/op
HeadersBenchmark.httpAddAllFastest              FIVE  avgt   10    97.299 ±   7.988  ns/op
HeadersBenchmark.httpAddAllFast                 FIVE  avgt   10   238.553 ±   2.578  ns/op
HeadersBenchmark.httpAddAllSlow                 FIVE  avgt   10   233.353 ±  18.746  ns/op
HeadersBenchmark.httpAddAllFastest         TWENTYTWO  avgt   10   397.501 ±  32.275  ns/op
HeadersBenchmark.httpAddAllFast            TWENTYTWO  avgt   10  1705.913 ± 269.910  ns/op
HeadersBenchmark.httpAddAllSlow            TWENTYTWO  avgt   10  1616.167 ±  47.053  ns/op
```

Note that I switched to using 10 iterations & 10 warmup iterations with 1 fork. Results seemed more stable that way

**Comment by Scottmitch** [comment]
@louiscryan - Thanks for the updated results!

@buchgr - Are you still looking into the difference between HTTP and HTTP/2 for the `Fastest` code path? The HTTP/2 headers do have a difference from HTTP headers in the `add0` method. For HTTP/2 the `newHeaderEntry` method is overridden to enforce the pseudo headers first iteration order.  I wonder if this is disabled (just for the benchmark) how that would impact the benchmark results....let me check this.

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java (hunk 5 above)
Comment (louiscryan) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: Done here and a few other similar places.

**Comment by Scottmitch** [comment]
@buchgr - My previous hunch seemed to have some impact. After commenting out `newHeaderEntry` and `Http2HeaderEntry` in `DefaultHttp2Headers` I get the following results:

Header size of 8:
```
Benchmark                            Mode  Cnt    Score   Error  Units
HeadersBenchmark.http2AddAllFastest  avgt    5  102.395 ± 7.644  ns/op
HeadersBenchmark.httpAddAllFastest   avgt    5  101.258 ± 4.519  ns/op
```

**Comment by louiscryan** [comment]
I believe the difference is because HTTP2 has to order pseduo headers before non-pseudo headers and this incurs the overhead

**Comment by louiscryan** [comment]
... and I'm too slow on the draw again :)

**Comment by Scottmitch** [comment]
@louiscryan :wink: We can look into optimizing this pseudo header ordering enforcement as a followup PR if necessary.

File: microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java [hunk 12]
```diff
@@ -33,43 +35,64 @@
...
-    @Param
+    private static String toHttpName(String name) {
+        return (name.startsWith(":")) ? name.substring(1) : name;
+    }
+
+    private static String toHttp2Name(String name) {
+        name = name.toLowerCase();
+        return (name.equals("host")) ? "xhost" : name;
+    }
+
```
Comment (Scottmitch) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: I think we need the `@Param` on this or else NPE?

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java [hunk 13]
```diff
@@ -64,6 +66,50 @@ public CharSequence escape(CharSequence value) {
...
     @Override
+    public HttpHeaders add(Headers<? extends CharSequence, ? extends CharSequence, ?> headers) {
+        // Override the fast-copy mechanism used by DefaultHeaders
+        if (headers == this) {
+            throw new IllegalArgumentException("can't add to itself.");
+        }
+        if (headers instanceof CombinedHttpHeaders) {
+            if (isEmpty()) {
+                // Can use the fast underlying copy
+                super.addImpl(headers);
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: nit: remove `super`?

**Comment by Scottmitch** [comment]
@louiscryan - Few more comments then I think we are good.

File: microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java (hunk 12 above)
Comment (louiscryan) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: Yup, fixed.

File: codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java (hunk 13 above)
Comment (louiscryan) [inline codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java]: Done

**Comment by louiscryan** [comment]
@Scottmitch Should be good to go now

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java [hunk 14]
```diff
@@ -76,8 +76,8 @@ public void testAddSelf() {
         headers.add(headers);
     }
 
-    @Test(expected = IllegalArgumentException.class)
-    public void testSetSelf() {
+    @Test
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java]: Does this mean we change the semantic ? It seems so to me...

File: codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java [hunk 15]
```diff
@@ -390,8 +443,8 @@ public void testAddSelf() {
         headers.add(headers);
     }
 
-    @Test(expected = IllegalArgumentException.class)
-    public void testSetSelf() {
+    @Test
+    public void testSetSelfIsNoOp() {
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/DefaultHeadersTest.java]: same comment as above

File: microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java [hunk 16]
```diff
@@ -139,4 +163,546 @@ public void http2Iterate(Blackhole bh) {
             bh.consume(entry);
         }
     }
+
+    @Benchmark
+    @BenchmarkMode(Mode.AverageTime)
+    public void httpAddAllFastest(Blackhole bh) {
+        bh.consume(new DefaultHttpHeaders(false).add(httpHeaders));
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: I think we should create the DefaultHttpHeaders as part of setup as otherwise these will also influence the benchmark. Same is true for other methods.

Comment (buchgr) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: That's a good point. That would require to clear the headers before each invocation though. Something like

```java
@Setup(Level.Invocation)
public void setup() {
   addAllFastestHeaders.clear();
}

@Benchmark
public void httpAddAllFastest() {
 addAllFastestHeaders.add(httpHeaders);  
}

```

Would have to test how this works out. Allocations would be less but not much less, as we mostly allocate `HeaderEntry` objects and those would also be there in `clear()`.

Adding `-verbose:gc` to the JVM flags, opens one's eyes to much GC is going on during a benchmark.

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java (hunk 14 above)
Comment (Scottmitch) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java]: I think it does, and we discussed this recently, but I don't think this violates the API.  I'm still in favor of only throwing an exception if API semantics can't be preserved.

Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java]: I would prefer we keep the current semantic. If we want to change it we should discuss and ensure it is the same in 4.0, 4.1 and master

File: microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java (hunk 16 above)
Comment (louiscryan) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: Happy to make the change though I agree it will have marginal impact on the actual numbers.

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java (hunk 14 above)
Comment (louiscryan) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java]: This is definitely a change in semantic and so worthy of discussion. 

I think the problem in the original code is the inconsistency in behavior between this.setAll(this) and this.set(this), one would throw and the other would not.
- these APIs are relatively easy for users to confuse 
- if the preconditions were dropped they would have the same outcome, they are both idempotent for this case
- the behavior was not documented

I suspect exposure to the semantic difference is rare, it would be odd to see this occur and odder still to be depending on it throwing for some real conditional behavior

File: microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java (hunk 16 above)
Comment (louiscryan) [inline microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java]: So the law of unintended consequences kicked in....

There was a memory leak in DefaultHttp2Headers.clear as it didn't reset the pseudoHeader pointer to head and retained a reference to the data from the last run. This caused the benchmarks to go haywire as you might imagine. Once I fixed that the numbers were about equivalent to what I got before.

```
Benchmark                            (exampleHeader)  Mode  Cnt     Score     Error  Units
HeadersBenchmark.http2AddAllFastest             FIVE  avgt   10   136.940 ±   3.899  ns/op
HeadersBenchmark.http2AddAllFast                FIVE  avgt   10   199.410 ±   3.022  ns/op
HeadersBenchmark.http2AddAllSlow                FIVE  avgt   10   231.186 ±   8.891  ns/op
HeadersBenchmark.http2AddAllFastest        TWENTYTWO  avgt   10   479.186 ±   6.863  ns/op
HeadersBenchmark.http2AddAllFast           TWENTYTWO  avgt   10  1014.660 ±  13.290  ns/op
HeadersBenchmark.http2AddAllSlow           TWENTYTWO  avgt   10   971.376 ±  12.329  ns/op
HeadersBenchmark.httpAddAllFastest              FIVE  avgt   10   100.142 ±   2.534  ns/op
HeadersBenchmark.httpAddAllFast                 FIVE  avgt   10   244.474 ±   3.014  ns/op
HeadersBenchmark.httpAddAllSlow                 FIVE  avgt   10   233.950 ±  33.473  ns/op
HeadersBenchmark.httpAddAllFastest         TWENTYTWO  avgt   10   415.668 ±  24.973  ns/op
HeadersBenchmark.httpAddAllFast            TWENTYTWO  avgt   10  1743.317 ±  13.812  ns/op
HeadersBenchmark.httpAddAllSlow            TWENTYTWO  avgt   10  1673.307 ± 125.075  ns/op
```

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java (hunk 14 above)
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java]: I agree and I'm fine with your change as long as you also back port it to 4.0 ;)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java [hunk 17]
```diff
@@ -146,6 +146,12 @@ public CharSequence status() {
     }
 
     @Override
+    public Http2Headers clear() {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java]: I wonder if we do an extra pr for this as this is not really related but a bug that you fixed. If we do an extra pr it is easier to keep track of changes for the user and see if an issue is fixed for them.

Comment (louiscryan) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java]: Happy to do that though I will have to revert the benchmarks to stop using 'clear' and instead create a new object each time to copy into.

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java (hunk 14 above)
Comment (louiscryan) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpHeadersTest.java]: So looking at the 4.0 HttpHeader class I think whats in this PR is more compatible with that than what was there before.

If 4.0  headers.set(headers) would not throw but it was semantically broken as it would do

this.clear()
<add all entries>

which would have the unintended consequence of emptying the headers instead of leaving them unchanged. In short there's a nasty bug in 4.0 rather than a semantic incompatability.

I'm happy to create a separate PR to fix the issue in 4.0.

There is no setAll in 4.0

**Comment by louiscryan** [comment]
Working on PR splitting now... Please hold

**Comment by louiscryan** [comment]
OK. Should be ready to go now that the other PRs are done.

**Comment by normanmaurer** [comment]
@louiscryan LGTM

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (6e108cb) master (1cdf0cb)

@louiscryan - Thanks! Make sure you follow the commit message template next time.
